x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -311,6 +331,48 @@ private void getVersionedDocument(String uid, ActionListener<VersionedDocument>
         });
     }
 
+    private void getVersionedDocuments(Collection<String> uids, ActionListener<Tuple<List<ProfileDocument>, List<String>>> listener) {
+        if (uids.isEmpty()) {
+            listener.onResponse(new Tuple<>(List.of(), List.of()));
+            return;
+        }
+        tryFreezeAndCheckIndex(listener).ifPresent(frozenProfileIndex -> {
+            frozenProfileIndex.checkIndexVersionThenExecute(
+                listener::onFailure,
+                () -> new OriginSettingClient(client, SECURITY_PROFILE_ORIGIN).prepareMultiGet()
+                    .addIds(frozenProfileIndex.aliasName(), uids.stream().map(ProfileService::uidToDocId).toArray(String[]::new))
+                    .execute(ActionListener.wrap(multiGetResponse -> {
+                        List<ProfileDocument> retrievedDocs = new ArrayList<>(multiGetResponse.getResponses().length);
+                        List<String> failures = new ArrayList<>(0);
+                        Exception loggedException = null;
+                        for (MultiGetItemResponse itemResponse : multiGetResponse.getResponses()) {
+                            if (itemResponse.isFailed() && itemResponse.getFailure() != null) {
```

Comment:
Internally `isFailed()` is just `return failure != null;` and `getFailure()` is `return this.failure;`. So I think just check `isFailed()` is enough here. Also other branches can be simplified as well because based on `MultiGetItemResponse`'s logic, one of `failure` and `response` must be non-null and we can directly call `MultiGetItemResponse#getId()` which internally delegates to either failure or response.

Reply:
I was well aware of the internal logic of the mget response that you're describing.
This is why I've put a catch-all else with
```
 logger.error("Inconsistent get item response [{}] [{}]", itemResponse.getIndex(), itemResponse.getId());
```

My approach was to be defensive wrt to inconsistent broken mget results.
But based on your feedback I have pushed 516784e8f2e175de374c6f16d1 . Hopefully you count it as an improvement.